src/librustc_jobserver/lib.rs

```diff
@@ -0,0 +1,292 @@
+//! acquire/release requests fall through to the jobserver crate's
+//! acquire/release functions).
+//!
+//! This can be quite poor for scalability, as at least on current Linux
+//! kernels, each release on the jobserver will trigger the kernel to wake up
+//! *all* waiters instead of just one, which, if you have lots of threads
+//! waiting, is quite bad.
+//!
+//! For that reason, we have a second mode which utilizes Cargo to improve
+//! scaling here. In that mode, we have slightly altered communication with the
+//! jobserver. Instead of just blocking on the jobserver, we will instead first
+//! print to stderr a JSON message indicating that we're interested in receiving
+//! a jobserver token, and only then block on actually receiving said token. On
+//! release, we don't write into the jobserver at all, instead merely printing
+//! out that we've released a token.
+//!
+//! Note that the second mode allows Cargo to hook up each rustc with its own
+//! jobserver (i.e., one per rustc process) and then fairly easily make sure to
+//! fulfill the requests from rustc and such. Ultimately, that means that we
+//! have just one rustc thread waiting on the jobserver: a solution that is
+//! nearly optimal for scalability.
+
+use jobserver_crate::Client;
+use lazy_static::lazy_static;
+use rustc_serialize::json::as_json;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::sync::Condvar;
+use std::sync::Mutex;
+
+lazy_static! {
+    // We can only call `from_env` once per process
+
+    // Note that this is unsafe because it may misinterpret file descriptors
+    // on Unix as jobserver file descriptors. We hopefully execute this near
+    // the beginning of the process though to ensure we don't get false
+    // positives, or in other words we try to execute this before we open
+    // any file descriptors ourselves.
+    //
+    // Pick a "reasonable maximum" if we don't otherwise have
+    // a jobserver in our environment, capping out at 32 so we
+    // don't take everything down by hogging the process run queue.
+    // The fixed number is used to have deterministic compilation
+    // across machines.
+    //
+    // Also note that we stick this in a global because there could be
+    // multiple rustc instances in this process, and the jobserver is
+    // per-process.
+    static ref GLOBAL_CLIENT: Client = unsafe {
+        Client::from_env().unwrap_or_else(|| {
+            log::trace!("initializing fresh jobserver (not from env)");
+            let client = Client::new(32).expect("failed to create jobserver");
+            // Acquire a token for the main thread which we can release later
+            client.acquire_raw().ok();
+            client
+        })
+    };
+
+    static ref HELPER: Helper = Helper::new();
+}
+
+const EMPTY: usize = 0;
+const CARGO_REQUESTED: usize = 1;
+const MAKE_REQUESTED: usize = 2;
+static TOKEN_REQUESTS: AtomicUsize = AtomicUsize::new(0);
```

Comment:
Could this be documented as to what it's used for briefly? (along with the values)